proposals/0303-swiftpm-extensible-build-tools.md

```diff
@@ -432,77 +470,106 @@ protocol Path: ExpressibleByStringLiteral, CustomStringConvertible {
+Because they run on every build, prebuild commands should use caching to do as little work as possible (ideally none) when there are no changes to their inputs. This should include preserving timestamps of generated source files that haven't changed, since most build systems (including SwiftPM's own build system) use timestamps as a shorthand for detecting whether files have changed.
+
+##### Build Commands
+
+Commands created using `addBuildCommand()` are incorporated into the build system's dependency graph, so that they run as needed during the build, based on their declared inputs and outputs. This requires that the paths of any outputs can be known before the command is run. This is usually done by forming the names of the outputs from some combination of the output directory and the name of the input file.
+
+Examples of plugins that can use regular build commands include compiler-like translators such as Protobuf and other tools that take a fixed set of inputs and produce a fixed set of outputs. (note that one nuance with Protobuf in particular is that it is actually up to the source generator invoked by `protoc` to determine the output paths — however, the relevant source generators for Swift and C do produce output files with predictable names).
+
+Regular build commands with defined outputs are preferable whenever possible, because such commands don't have to run unless their outputs are missing or their inputs have changed since the last time they ran.
+
+##### Postbuild Commands
+
+Commands created using `addPostbuildCommand()` are run after every build. They have no input or output directories, since they provide no inputs to the build.
+
+Instead, they can be passed information about whether the build succeeded or failed, along with other information about the build. SwiftPM vends this information via a JSON file whose path is provided in the `postbuildResultFile` property of the target build context passed to the plugin. The plugin may choose to pass this path to the postbuild command in any way that it wants to, such as on the command line or in an environment variable.
+
+The postbuild result file is only guaranteed to actually be written if at least one postbuild command is defined. It will not exist while prebuild and regular build commands are running.
 
-#### Error Handling
+The format of the JSON file written at the end of the build is initially very minimal, and consists only of these keys:
+
+- `succeeded` — a boolean that is true if and only if the command completed successfully (note that if the build was cancelled, the postbuild command is not run at all)
+- `buildConfiguration` — the build configuration that was built (either `debug` or `release`)
+- `productDirectory` — absolute path of the directory containing the built products
```

Comment:
I see, yeah that's minimal enough :) 

Nice to have debug/release.

No need for more in initial version I think

Reply:
A question which may come up: so... would a postBuild be a source formatter or it can't get access to the sources?

Reply:
It can read the sources but not write to them.  I realize that many source formatters today rely on being able to modify the sources, but for security reasons I don't think we want to allow that in general.  We should extend the API to allow them to create a structured diagnostic file that has fixits, and then make it easy for the end user to view and apply those fixits.

Reply:
> but for security reasons I don't think we want to allow that in general

can you elaborate on that a bit?

Reply:
> A question which may come up: so... would a postBuild be a source formatter or it can't get access to the sources?

could/should it be preBuild? since you want the build to use/validate the new formatted code?

Reply:
> We should extend the API to allow them to create a structured diagnostic file that has fixits, and then make it easy for the end user to view and apply those fixits.

I think this is a great idea for supporting linters, feel like a new plugin "type" or "capability"

Reply:
> > but for security reasons I don't think we want to allow that in general
> 
> can you elaborate on that a bit?

It seems risky to allow a plugin to have write access to the source directory during a regular build, since it would be easy for it to make changes that the package author isn't aware of.  What would make sense would be to allow this as maybe a new capability that adds a subcommand to `swift package`, where the whole point is on-demand modification of the source.  But if I just apply a plugin and do `swift build`, it seems prudent for the sandbox to prevent modification of the sources.

Reply:
> > We should extend the API to allow them to create a structured diagnostic file that has fixits, and then make it easy for the end user to view and apply those fixits.
> 
> I think this is a great idea for supporting linters, feel like a new plugin "type" or "capability"

We should have a new capability for linters and code formatters, but I do think that allowing regular build tools to pass along structured diagnostics would be good in general.  For example, `protoc` should ideally be able to emit fixits, etc.  That's something we can add over time, since we'd also need to have a way for those tools to created the structured diagnostics.

Reply:
> > > but for security reasons I don't think we want to allow that in general
> > 
> > 
> > can you elaborate on that a bit?
> 
> It seems risky to allow a plugin to have write access to the source directory during a regular build, since it would be easy for it to make changes that the package author isn't aware of. What would make sense would be to allow this as maybe a new capability that adds a subcommand to `swift package`, where the whole point is on-demand modification of the source. But if I just apply a plugin and do `swift build`, it seems prudent for the sandbox to prevent modification of the sources.

Although I suppose in practice there isn't a large distinction between having a tool that generates new code (which can contain anything) vs modifying existing code.  Perhaps it's more that it seems surprising to do a build and then end up with modification to the code as a side effect.

Reply:
> What would make sense would be to allow this as maybe a new capability that adds a subcommand to swift package, where the whole point is on-demand modification of the source.

I think this is a capability / plugin type we need to add regardless of if those perform source modifications, but not sure this needs to be part of the initial API.

Reply:
> > What would make sense would be to allow this as maybe a new capability that adds a subcommand to swift package, where the whole point is on-demand modification of the source.
> 
> I think this is a capability / plugin type we need to add regardless if it perform source modifications, but not sure this needs to be part of the initial API.

Absolutely agreed.  I was just suggesting the direction that should take, not that it should be added here.